sim.c

```diff
@@ -615,8 +615,28 @@ static void enqueue_movement(struct board *board, struct movement m)
     board->movements.movements[board->movements.length++] = m;
 }
 
-static void move_atom(struct board *board, atom *a, struct vector position)
+static bool movements_equal(struct movement *a, struct movement *b)
 {
+    if (a->rotation == 0 && b->rotation == 0)
+        return vectors_equal(a->translation, b->translation);
+    if (a->rotation != b->rotation)
+        return false;
+    struct vector a_center = a->type == SWING_MOVEMENT ? a->base : a->absolute_grab_position;
+    struct vector b_center = b->type == SWING_MOVEMENT ? b->base : b->absolute_grab_position;
+    return vectors_equal(a_center, b_center);
+}
+
+static void move_atom(struct board *board, atom *a, struct vector position, struct movement *movement)
+{
+    if (*a & BEING_MOVED) {
+        size_t index = *a & MOVEMENT_INDEX;
+        if (index > board->movements.length)
+            report_collision(board, position, "this should never happen");
```

Comment:
For "this should never happen" cases, I've been using `abort()` (maybe this is a bit rude for a library to do, though).  If this is reported as a collision, I think it would at least be good to write a descriptive message like "atom movement index exceeded length of movements list" so we have an idea of what happened if it actually does happen.